components/raftstore/src/store/worker/disk_check.rs

```diff
@@ -0,0 +1,139 @@
+// Copyright 2024 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{
+    fmt::{self, Display, Formatter},
+    io::Write,
+    path::PathBuf,
+    time::Duration,
+};
+
+use health_controller::types::LatencyInspector;
+use tikv_util::{
+    time::Instant,
+    warn,
+    worker::{Builder as WorkerBuilder, Runnable, Worker},
+};
+
+#[inline]
+pub fn init_disk_check_worker() -> Worker {
+    // The disk check mechanism only cares about the latency of the most
+    // recent request; older requests become stale and irrelevant. To avoid
+    // unnecessary accumulation of multiple requests, we set a small
+    // `pending_capacity` for the disk check worker.
+    WorkerBuilder::new("disk-check-worker")
+        .pending_capacity(3)
+        .create()
+}
+
+/// A simple inspector to measure the latency of disk IO.
+///
+/// This is used to measure the latency of disk IO, which is used to determine
+/// the health status of the TiKV server.
+/// The inspector writes a file to the disk and measures the time it takes to
+/// complete the write operation.
+pub struct Runner {
+    target: PathBuf,
+}
+
+impl Runner {
+    /// The filename to write to the disk to measure the latency.
+    const DISK_IO_LATENCY_INSPECT_FILENAME: &'static str = ".disk_latency_inspector.tmp";
+    /// The content to write to the file to measure the latency.
+    const DISK_IO_LATENCY_INSPECT_FLUSH_STR: &'static [u8] = b"inspect disk io latency";
+
+    #[inline]
+    pub fn new(inspect_dir: PathBuf) -> Self {
+        Self {
+            target: inspect_dir.join(Self::DISK_IO_LATENCY_INSPECT_FILENAME),
+        }
+    }
+
+    /// Only for test.
+    /// Generate a dummy Runner.
+    pub fn dummy() -> Self {
+        Self {
+            target: PathBuf::from("./").join(Self::DISK_IO_LATENCY_INSPECT_FILENAME),
+        }
+    }
+
+    fn inspect(&self) -> Option<Duration> {
+        let mut file = std::fs::OpenOptions::new()
+            .create(true)
+            .write(true)
+            .truncate(true)
+            .open(&self.target)
+            .ok()?;
+
+        let start = Instant::now();
+        // Ignore the error
+        file.write_all(Self::DISK_IO_LATENCY_INSPECT_FLUSH_STR)
+            .ok()?;
+        file.sync_all().ok()?;
+        Some(start.saturating_elapsed())
+    }
+}
+
+#[derive(Debug)]
+pub enum Task {
+    InspectLatency { inspector: LatencyInspector },
+}
+
+impl Display for Task {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match *self {
+            Task::InspectLatency { .. } => write!(f, "InspectLatency"),
+        }
+    }
+}
+
+impl Runnable for Runner {
```

Comment:
You can use `RunnableWithTimer`. Runner can periodically measure disk latency `on_timeout`.

Reply:
IMO, not recommended by using `RunnableWithTimer `.

The reason why we choose a single worker to do this, including several aspects:
- It cannot cover the scenario that if the kvdb is fully stuck or hung on Disk I/O (unless it uses `async` function with timeout to do the check).
- Follows current implementation on inspecting raft disk I/O latency. This runner is expected to be a individual receiver and executor to call relative function to detect the inspecting tasks. The inspector in pd-worker will trigger a task periodically and sending it to this runner. And if the runner cannot returns the response by calling the callback, pd-worker can catch the timeout cases as expected and timely updates the relative SlowScore.

Reply:
I think it's possible to use `RunnableWithTimer`, but it adds a little more complexity and I'm not sure how much value it adds over the current version.

With `RunnableWithTimer`, the runner can record disk check results internally and the inspect latency request can query those results on each slow score tick. The fully stuck scenario will be detected by an inspect request timeout, similar to how it is detected now. One benefit might be that the disk check interval and timeout can be decoupled from the inspect request interval.